src/termination.ml

```diff
@@ -0,0 +1,378 @@
+         let c = ref 0 in
+         List.iter (incr c; print_tl_rule oc !c s) !(s.sym_rules)
+    ) else ()
+  in
+  (* Print the variable declarations *)
+  let print_vars : sym -> unit = fun s ->
+    let c = ref 0 in
+    List.iter
+      (fun r ->
+        incr c;
+        List.iter
+          (fun (x,ty) ->
+            Format.fprintf oc
+              "<varDeclaration>@.<var>%s_%i_%s</var>@." s.sym_name !c x;
+            Format.fprintf oc
+              "<type>@.%a</type>@.</varDeclaration>@."
+              (print_type !c s.sym_name) ty
+          )
+          (get_vars s r)
+      )
+      !(s.sym_rules)
+  in
+  (* Print the symbol declarations at object level. *)
+  let print_symbol : sym -> unit = fun s ->
+    if status s = Object_level
+    then (
+      Format.fprintf oc "<funcDeclaration>@.<name>%a</name>@." print_sym s;
+      Format.fprintf oc
+        "<typeDeclaration>@.<type>@.%a</type>@."
+        (print_type 0 s.sym_name) !(s.sym_type);
+      Format.fprintf oc "</typeDeclaration>@.</funcDeclaration>@."
+    )
+  in
+  (* Print the type constructor declarations. *)
+  let print_type_cstr : sym -> unit = fun s ->
+    (* We don't declare type constant which do not take arguments for
+       compatibility issue with simply-typed higher order category of the
+       competition. *)
+    if status s = Type_cstr
+    then (
+      Format.fprintf oc "<typeConstructorDeclaration>@.<name>%a</name>@."
+        print_sym s;
+      Format.fprintf oc
+        "<typeDeclaration>@.<type>@.%a</type>@." (print_type 0 s.sym_name) !(s.sym_type);
+      Format.fprintf oc "</typeDeclaration>@.</typeConstructorDeclaration>@."
+    )
+  in
+  List.iter (Format.fprintf oc "%s@.") prelude;
+  Format.fprintf oc "<rules>@.";
+  iter_symbols print_object_rules;
+  Format.fprintf oc "</rules>@.";
+  let type_rule_presence = ref false in
+  iter_symbols
+    (fun s ->
+      if status s != Object_level && !(s.sym_def) != None
+         && !(s.sym_rules) != []
+      then type_rule_presence := true);
+  if !type_rule_presence
+  then (
+    Format.fprintf oc "<typeLevelRules>@.";
+    iter_symbols print_type_rules;
+    Format.fprintf oc "</typeLevelRules>@."
+  );
+  Format.fprintf oc "<higherOrderSignature>@.";
+  Format.fprintf oc "<variableTypeInfo>@.";
+  iter_symbols print_vars;
+  Format.fprintf oc "</variableTypeInfo>@.";
+  Format.fprintf oc "<functionSymbolTypeInfo>@.";
+  iter_symbols print_symbol;
+  Format.fprintf oc "</functionSymbolTypeInfo>@.";
+  let type_cstr_presence = ref false in
+  iter_symbols
+    (fun s -> if status s = Type_cstr then type_cstr_presence := true);
+  if !type_cstr_presence
+  then (
+    Format.fprintf oc "<typeConstructorTypeInfo>@.";
+    iter_symbols print_type_cstr;
+    Format.fprintf oc "</typeConstructorTypeInfo>@."
+  );
+  Format.fprintf oc "</higherOrderSignature>@.";
+  List.iter (Format.fprintf oc "%s@.") postlude
+  
+(** [check cmd sign] runs the termination checker specified by command [cmd] on
```

Comment:
Why duplicating this code from confluence.ml? This should have been factorized!

Reply:
This PR has been merged too quickly...

Reply:
I disagree, this PR remained open long enough. Moreover, it only have a very small impact on the rest of the code, the new feature being quite orthogonal. Improvements can be brought by subsequent PR(s).

Reply:
Note that I did not spend much time reading `src/termination.ml`, since it does not have any impact on current features of the prototype.